### PR TITLE
Default to first non-disabled option for select elements

### PR DESCRIPTION
**Pull request by aweary**
This fixes a regression that @nhunzaker found in https://github.com/facebook/react/pull/10142 where `select` elements were defaulting to the first `option`, regardless of whether it was disabled or not. This was not the behavior in 15.6. Copying my comment on this from Nate's PR:

>This issue stems from https://github.com/facebook/react/pull/8349 which introduced an `updateOptions` call when a `select` element is mounted.  If`updateOptions` doesn't find a matching `option` it [defaults to setting the first `option` as selected](https://github.com/facebook/react/blob/ca46c5278fe24dd9f564918d30e454033278e054/src/renderers/dom/fiber/wrappers/ReactDOMFiberSelect.js#L110-L112).
>
>In previous versions `updateOptions` was not called on the initial mount, so it defaulted to the standard behavior of the browser.

This brings us back in sync with the HTML spec which states:

> If the select element's display size is 1, and no option elements in the select element's list of options have their selectedness set to true, set the selectedness of the first option element in the list of options in tree order **that is not disabled**, if any, to true.

https://html.spec.whatwg.org/multipage/form-elements.html#the-select-element

@nhunzaker I can just cherry-pick the DOM fixtures from #10142 as part of this PR if you'd like!

### CODE REVIEW

File: src/renderers/dom/fiber/wrappers/ReactDOMFiberSelect.js [hunk 1]
```diff
@@ -101,14 +101,18 @@ function updateOptions(
...
         return;
       }
+      if (defaultSelected === null && !options[i].disabled) {
+        defaultSelected = options[i];
+      }
     }
-    if (options.length) {
-      options[0].selected = true;
+    if (defaultSelected !== null) {
+      defaultSelected.selected = true;
```
Comment (jquense) [inline src/renderers/dom/fiber/wrappers/ReactDOMFiberSelect.js]: I know you didn't write all this, but the logic is hard to follow here...can we switch early to see if it's controlled?

Comment (aweary) [inline src/renderers/dom/fiber/wrappers/ReactDOMFiberSelect.js]: I agree it's a little dense, can you clarify what you mean by switching early to see if it's controlled?  As far as I can tell, `updateOptions` only branches on `multiple` and doesn't really care if it's controlled or not.

**Comment by sebmarkbage** [comment]
What happens to browser features like attempts to automatically restoring state and auto-fill? Does it care?

**Comment by nhunzaker** [comment]
@aweary 🍒 ⛏ away.

**Comment by nhunzaker** [comment]
@sebmarkbage We don't have fixture coverage for restored state and auto-fill. This would be a good addition to the suite. I'll write up an issue.

File: src/renderers/dom/fiber/wrappers/ReactDOMFiberSelect.js [hunk 2]
```diff
@@ -101,14 +101,18 @@ function updateOptions(
     // Do not set `select.value` as exact behavior isn't consistent across all
     // browsers for all cases.
```
Comment (nhunzaker) [inline src/renderers/dom/fiber/wrappers/ReactDOMFiberSelect.js]: I wish I knew _what_ wasn't consistent -- and where? Like was it IE8?. I'll write up an issue to do some archaeology here. 

Totally unrelated to this PR 😛.

File: src/renderers/dom/fiber/wrappers/ReactDOMFiberSelect.js (hunk 1 above)
Comment (jquense) [inline src/renderers/dom/fiber/wrappers/ReactDOMFiberSelect.js]: I was thinking check if `propValue` was set, but that wouldn't actually work...I guess i don't have any specific suggestions :P but I felt like I had to read this through 3 times to understand it, and it _feels_ like it should be easier to express

**Comment by gaearon** [comment]
What is the status on this?

**Comment by aweary** [comment]
@gaeron I cherry-picked the test fixtures, so it should be good. Feel free to merge if it looks good to you.

**Comment by gaearon** [comment]
The fixture is broken for me.
There's a few `const`s before `import`s (probably because `react-scripts` version has changed since).

But I also get a runtime error:

screen shot 2017-09-01 at 1 14 51 pm

Inside `selects/index.js`, `_nestedDOMNode` is always `null` so `_renderNestedSelect` fails.

File: fixtures/dom/src/components/fixtures/selects/index.js [hunk 3]
```diff
@@ -31,35 +34,66 @@ class SelectFixture extends React.Component {
...
-          <select defaultValue="">
-            <option value="">Select a color</option>
-            <option value="red">Red</option>
-            <option value="blue">Blue</option>
-            <option value="green">Green</option>
-          </select>
-          <span className="hint" />
-        </fieldset>
-        <fieldset>
-          <legend>Controlled in nested subtree</legend>
```
Comment (gaearon) [inline fixtures/dom/src/components/fixtures/selects/index.js]: Was this case removed intentionally?

**Comment by gaearon** [comment]
Thanks!

**Comment by aweary** [comment]
Thanks for the fix @gaearon!
